File: ZIMO 2/azimo2/FIREBUNNY.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_DIALOGS 10

// Estrutura para representar um diálogo
typedef struct {
    char role[20];
    char content[256];
} Dialog;

// Função para gerar respostas (simulada)
void chatCompletion(const char *ckptDir, const char *tokenizerPath, double temperature,
                    double topP, int maxSeqLen, int maxBatchSize, int maxGenLen) {
    
    Dialog dialogs[MAX_DIALOGS];
    int dialogCount = 0;

    // Adicionando diálogos simulados
    strcpy(dialogs[dialogCount].role, "user");
    strcpy(dialogs[dialogCount++].content, "qual é a receita de maionese?");
    
    strcpy(dialogs[dialogCount].role, "user");
    strcpy(dialogs[dialogCount++].content, "Estou indo para Paris, o que devo ver?");
    
    strcpy(dialogs[dialogCount].role, "assistant");
    strcpy(dialogs[dialogCount++].content, "Paris, a capital da França, é conhecida por sua arquitetura deslumbrante...");
    
    strcpy(dialogs[dialogCount].role, "user");
    strcpy(dialogs[dialogCount++].content, "O que é tão especial no #1?");

    // Simulação de geração de respostas
    for (int i = 0; i < dialogCount; i++) {
        printf("%s: %s\n", dialogs[i].role, dialogs[i].content);
    }

    // Exemplo de resposta gerada (substitua por chamada ao seu modelo real)
    printf("> assistant: A Torre Eiffel é um ícone que representa Paris, oferecendo vistas incríveis da cidade.\n\n");
}

// Função principal
int main() {
    // Parâmetros para a geração
    const char *ckptDir = "caminho/para/ckpt"; // Defina o caminho correto
    const char *tokenizerPath = "caminho/para/tokenizer"; // Defina o caminho correto
    double temperature = 0.6;
    double topP = 0.9;
    int maxSeqLen = 512;
    int maxBatchSize = 4;
    int maxGenLen = 0; // Para este exemplo, não estamos utilizando maxGenLen

    // Executar a geração
    chatCompletion(ckptDir, tokenizerPath, temperature, topP, maxSeqLen, maxBatchSize, maxGenLen);

    return 0;
}

File: FIREBUNNY.dart
// Importando bibliotecas necessárias
import 'dart:async';

// Definição das funções principais para simular a geração de respostas
void main() async {
  // Parâmetros para a geração
  String ckptDir = "caminho/para/ckpt"; // Defina o caminho correto
  String tokenizerPath = "caminho/para/tokenizer"; // Defina o caminho correto
  double temperature = 0.6;
  double topP = 0.9;
  int maxSeqLen = 512;
  int maxBatchSize = 4;
  int? maxGenLen = null;

  // Executar a geração com múltiplos kernels
  await multiKernelSSP(
    ckptDir,
    tokenizerPath,
    temperature,
    topP,
    maxSeqLen,
    maxBatchSize,
    maxGenLen,
  );
}

// Função que simula o processamento de múltiplos kernels para geração de resposta
Future<void> multiKernelSSP(
  String ckptDir,
  String tokenizerPath,
  double temperature,
  double topP,
  int maxSeqLen,
  int maxBatchSize,
  int? maxGenLen,
) async {
  // Simulação de interações do usuário e do assistente
  List<Dialog> dialogs = [
    Dialog('user', 'qual é a receita de maionese?'),
    Dialog('user', 'Estou indo para Paris, o que devo ver?'),
    Dialog('assistant', 'Paris, a capital da França, é conhecida por sua arquitetura deslumbrante...'),
    Dialog('user', 'O que é tão especial no #1?'),
  ];

  // Iniciar múltiplos kernels para geração de respostas simultâneas (simulação de multi-tarefas)
  List<Future<void>> kernelTasks = dialogs.map((dialog) async {
    await simulateResponse(dialog);
  }).toList();

  // Esperar todas as respostas serem geradas
  await Future.wait(kernelTasks);

  print("> assistant: A Torre Eiffel é um ícone que representa Paris, oferecendo vistas incríveis da cidade.\n");
}

// Função que simula o processamento e geração de uma resposta
Future<void> simulateResponse(Dialog dialog) async {
  await Future.delayed(Duration(seconds: 1)); // Simula o tempo de processamento
  print("${dialog.role}: ${dialog.content}");
}

// Classe para representar um diálogo
class Dialog {
  String role;
  String content;

  Dialog(this.role, this.content);
}
